### PR TITLE
fix: apply resource conditions to DeploymentPreference resources

### PR DESCRIPTION
**Pull request by keetonian**
*Issue #, if available:*
N/A

*Description of changes:*
Bug fix. Adds conditions from the Serverless::Function resource to the CodeDeploy resources that are created when using `DeploymentPreference`. Previously, these resources were added even if the resources that use them were not included based on resource conditions.

*Description of how you validated changes:*
Updated tests, deployed sample template and verified that resources were created correctly.

* Breaking Change?
Potentially a breaking change if there is a dependency on the resources that were created before hand (when there was a bug, and conditions were not passed through to deployment preference resources). With this change, those resources that were previously created will be up for deletion. If those resources are being used elsewhere, those will start breaking now.

*Checklist:*

- [x] Write/update tests
- [x] `make pr` passes
- [ ] Update documentation
- [x] Verify transformed template deploys and application functions as expected
- [ ] Add/update example to `examples/2016-10-31`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by keetonian** [comment]
Deployed the following template (after a local transform) and verified that it created the correct resources (and didn't create others).

NOTE: This will cause some redeployments for customers where some resources will be removed from their templates. In the case of the following template:

**Resources created:**
+ CodeDeployServiceRole AWS::IAM::Role                                                                                          
+ MinimalFunctionAliaslive AWS::Lambda::Alias                                                                                      
+ MinimalFunctionDeploymentGroup AWS::CodeDeploy::DeploymentGroup 
+ MinimalFunctionRole AWS::IAM::Role                                                                                          
+ MinimalFunctionVersion76c37f65fb AWS::Lambda::Version 
+ MinimalFunction AWS::Lambda::Function                                                                                   
+ ServerlessDeploymentApplication AWS::CodeDeploy::Application    

**Resources no longer created (will be deleted):**
+ MinimalFunctionWithDeploymentPreferenceWithHooksAndAlarmsDeploymentGroup                                  AWS::CodeDeploy::DeploymentGroup 
+ MinimalFunctionWithMinimalDeploymentPreferenceDeploymentGroup AWS::CodeDeploy::DeploymentGroup

```yaml
Transform: AWS::Serverless-2016-10-31

Conditions:
  Condition1:
    Fn::Equals:
      - true
      - true
  Condition2:
    Fn::Equals:
      - true
      - false
  Condition3:
    Fn::Equals:
      - true
      - false

Parameters:
  MyFalseParameter:
    Type: String
    Default: False

Resources:
  MinimalFunction:
    Condition: Condition1
    Type: 'AWS::Serverless::Function'
    Properties:
      Handler: hello.handler
      InlineCode: |
        def hi():
          print("hi")
      Runtime: python2.7
      AutoPublishAlias: live
      DeploymentPreference:
        Type: Linear10PercentEvery2Minutes

  MinimalFunctionWithMinimalDeploymentPreference:
    Type: 'AWS::Serverless::Function'
    Condition: Condition2
    Properties:
      InlineCode: |
        def hi():
          print("hi")
      Handler: hello.handler
      Runtime: python2.7
      AutoPublishAlias: livewithdeployment
      DeploymentPreference:
        Type: Canary10Percent5Minutes

  MinimalFunctionWithDeploymentPreferenceWithHooksAndAlarms:
    Type: 'AWS::Serverless::Function'
    Condition: Condition2
    Properties:
      InlineCode: |
        def hi():
          print("hi")
      Handler: hello.handler
      Runtime: python2.7
      AutoPublishAlias: livewithdeploymentwithhooksandalarms
      DeploymentPreference:
        Type: Linear10PercentEvery2Minutes
        Hooks:
          PreTraffic: !Ref MySanityTestFunction
          PostTraffic: !Ref MyValidationTestFunction

  MySanityTestFunction:
    Type: 'AWS::Serverless::Function'
    Condition: Condition3
    Properties:
      Handler: hello.handler
      Runtime: python2.7
      InlineCode: |
        def hi():
          print("hi")
      DeploymentPreference:
        Enabled: False

  MyValidationTestFunction:
    Type: 'AWS::Serverless::Function'
    Condition: Condition3
    Properties:
      Handler: hello.handler
      Runtime: python2.7
      InlineCode: |
        def hi():
          print("hi")
      DeploymentPreference:
        Enabled: !Ref MyFalseParameter
```

**Comment by sriram-mv** [comment]
This PR potentially removes resources on a new deployment, the resources that will be removed should have **not** have been used. This PR does fixes an underlying bug and should have been the expected behavior to start with.

Any concerns on this PR?

**Comment by hawflau** [comment]
Pushed new changes to enable the condition fix only if the feature flag "deployment_preference_condition_fix" is enabled. Existing behavior will be used for any account ID that is not allow-listed (see test case "tests/translator/input/function_with_deployment_preference_condition_without_condition_fix.yaml"). 

As the above comment mentioned, the resources that will be removed should have not been used. We decided to use a feature flag to gate it before we are certain that removing the resources are safe. More investigation will be done but is not under the current scope

**Comment by jfuss** [comment]
@hawflau Can we just add the removal of these into a new Property? That way back compat is kept but customers have a way to remove the behavior as needed?

**Review by torresxb1 — APPROVED** [review]

File: samtranslator/model/preferences/deployment_preference.py [hunk 1]
```diff
@@ -75,8 +75,6 @@ def from_dict(cls, logical_id, deployment_preference_dict, condition=None):
         role = deployment_preference_dict.get("Role", None)
         trigger_configurations = deployment_preference_dict.get("TriggerConfigurations", None)
         passthrough_condition = deployment_preference_dict.get("PassthroughCondition", False)
-        # FIXME: We should support not only true/false, but also yes/no, on/off? See https://yaml.org/type/bool.html 
-        passthrough_condition = True if passthrough_condition in [True, "true", "True"] else False
```
Comment (torresxb1) [inline samtranslator/model/preferences/deployment_preference.py]: moved to `sam_resources`

**Review by hawflau — APPROVED** [review]

**Review by moelasmar — APPROVED** [review]

**Pull request merged**